### PR TITLE
Added activation functions to MLP models.

### LINKED ISSUE #43: Changes to network achitectures
For some reason the non-linear activation functions have been removed from the models in the solutions. This may well work, but it seems very irregular and, since this is for educational purposes, it should be corrected.

### PR DESCRIPTION
**Pull request by jdenholm**
- [x] Update the penguin classification demo by adding activation functions to the model.
- [x] Update the penguin regression demo.

Closes #43.

### CODE REVIEW

File: worked-solutions/01_penguin_classification_solutions.ipynb [hunk 1]
```diff
@@ -399,18 +400,20 @@
...
+      "    (9): Linear(in_features=16, out_features=3, bias=True)\n",
       "  )\n",
       ")\n"
      ]
     }
    ],
    "source": [
     "from torch.nn import Module\n",
-    "from torch.nn import BatchNorm1d, Linear, ReLU, Dropout, Sequential\n",
+    "from torch.nn import BatchNorm1d, Linear, LeakyReLU, Dropout, Sequential\n",
```
Comment (jatkinson1000) [inline worked-solutions/01_penguin_classification_solutions.ipynb:416]: Any reason to choose leaky over regular ReLU?
The network isn't very deep.
You're the expert - asking more for my own interest.

File: worked-solutions/01_penguin_classification_solutions.ipynb [hunk 2]
```diff
@@ -509,7 +514,7 @@
     "# Create an optimiser and give it the model's parameters.\n",
     "from torch.optim import Adam\n",
     "\n",
-    "optimiser = Adam(model.parameters())"
+    "optimiser = Adam(model.parameters(), lr=1e-4)"
```
Comment (jatkinson1000) [inline worked-solutions/01_penguin_classification_solutions.ipynb:517]: IIRC the choice of a learning rate is discussed in exercise 2, we left it as the default here so as not to over-complicate/increase the knowledge burden on exercise 1.

File: worked-solutions/01_penguin_classification_solutions.ipynb [hunk 3]
```diff
@@ -694,49 +699,29 @@
...
-      "Epoch 18 time: 0.096 seconds.\n",
-      "Epoch 19 time: 0.093 seconds.\n",
+      "Epoch 0-25 time: 4.040087 seconds\n",
+      "Epoch 25-50 time: 4.014584 seconds\n",
+      "Epoch 50-75 time: 4.016838 seconds\n",
+      "Epoch 75-100 time: 3.246220 seconds\n",
+      "Epoch 100-125 time: 2.165241 seconds\n",
+      "Epoch 125-150 time: 3.849997 seconds\n",
+      "Epoch 150-175 time: 3.943952 seconds\n",
+      "\n",
```
Comment (jatkinson1000) [inline worked-solutions/01_penguin_classification_solutions.ipynb:709]: This seems to take longer to train now, why?
Again, asking for my own interest rather than as a criticism.

File: worked-solutions/01_penguin_classification_solutions.ipynb [hunk 4]
```diff
@@ -850,8 +844,8 @@
       "        [3.3600e+01, 1.1300e+01, 2.0000e+03, 2.1100e+02, 1.0000e+00]])\n",
       "\n",
       "Raw output:\n",
-      "tensor([[0.0035, 0.0021, 0.9943],\n",
-      "        [0.9547, 0.0013, 0.0440]])\n",
+      "tensor([[0.0059, 0.0018, 0.9923],\n",
+      "        [0.5084, 0.0350, 0.4567]])\n",
```
Comment (jatkinson1000) [inline worked-solutions/01_penguin_classification_solutions.ipynb:848]: Predictive performance has worsened quite a bit on the second of these.

File: worked-solutions/02_penguin_regression_solutions.ipynb [hunk 5]
```diff
@@ -406,13 +406,7 @@
...
       "  )\n",
       "  (4): DenseBlock(\n",
-      "    (0): Linear(in_features=4, out_features=2, bias=True)\n",
-      "    (1): BatchNorm1d(2, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
-      "    (2): Dropout(p=0.1, inplace=False)\n",
-      "    (3): LeakyReLU(negative_slope=0.1)\n",
-      "  )\n",
-      "  (5): DenseBlock(\n",
-      "    (0): Linear(in_features=2, out_features=1, bias=True)\n",
+      "    (0): Linear(in_features=4, out_features=1, bias=True)\n",
```
Comment (jatkinson1000) [inline worked-solutions/02_penguin_regression_solutions.ipynb:409]: Why simplify this?
Again, asking to learn.

**Review by jatkinson1000 — COMMENTED** [review]
Generally looks good.
Couple of questions for my own understanding and couple of comments.
Once you look I'd be happy to merge.
Are you coming to the pub Thursday?

**Comment by jatkinson1000** [comment]
Hey @jdenholm is this ready to be merged do you think? It looks like the JOSE review is nearly passed so it would be good to merge this before that.

Aside from the point about dependencies I think the rest of my comments are purely questions for my own interest so no major objections.

File: worked-solutions/01_penguin_classification_solutions.ipynb (hunk 1 above)
Comment (jdenholm) [reply worked-solutions/01_penguin_classification_solutions.ipynb:416]: There's no set rule, but it can (sometimes) perform better. The only way to justify it really, though, is empirically.

File: worked-solutions/01_penguin_classification_solutions.ipynb (hunk 2 above)
Comment (jdenholm) [reply worked-solutions/01_penguin_classification_solutions.ipynb:517]: 1e-4 is super common with Adam, but so long as the model works it is fine to leave it as it was.

File: worked-solutions/01_penguin_classification_solutions.ipynb (hunk 3 above)
Comment (jdenholm) [reply worked-solutions/01_penguin_classification_solutions.ipynb:709]: Probably because the learning rate dropped. We could revert this change.

File: worked-solutions/01_penguin_classification_solutions.ipynb (hunk 4 above)
Comment (jdenholm) [reply worked-solutions/01_penguin_classification_solutions.ipynb:848]: Not sure that matters if the input isn't "real"? Happy to revert :)

**Comment by jdenholm** [comment]
I'm happy for you to change it as you please after my replies. Sorry for taking an age.

File: worked-solutions/01_penguin_classification_solutions.ipynb (hunk 1 above)
Comment (jatkinson1000) [reply worked-solutions/01_penguin_classification_solutions.ipynb:416]: Ah, I see it is now used in TorchTools, so makes sense to use here too. I also checked and we never explicitly (well, written) talk about ReLU or any other layer choice, so this could be left up to the person delivering the course to discuss.

**Review by jatkinson1000 — APPROVED** [review]
This is great, thanks for answering my questions @jdenholm.
I have addressed the minor points I made with a couple of follow-up commits and now all LGTM so will merge.
